File: backend/app/models/product.py
import datetime as dt

from sqlalchemy import Column, DateTime, Float, ForeignKey, Integer, String, Text
from sqlalchemy.orm import relationship

from app.database import Base


class ProductCategory(Base):
    __tablename__ = "product_categories"

    id = Column(Integer, primary_key=True, index=True)
    name = Column(String(150), nullable=False, unique=True)
    description = Column(Text, nullable=True)

    products = relationship("Product", back_populates="category")


class Product(Base):
    __tablename__ = "products"

    id = Column(Integer, primary_key=True, index=True)
    sku = Column(String(100), unique=True, index=True, nullable=False)
    name = Column(String(200), nullable=False)
    brand = Column(String(150), nullable=True)
    price = Column(Float, nullable=True)

    category_id = Column(Integer, ForeignKey("product_categories.id"), nullable=True)
    shelf_id = Column(Integer, ForeignKey("shelves.id"), nullable=True)

    # position of the product's facing on the shelf, JSON-encoded {x, y, w, h}
    shelf_position = Column(Text, nullable=True)
    image_url = Column(String(500), nullable=True)

    created_at = Column(DateTime, default=dt.datetime.utcnow)
    updated_at = Column(DateTime, default=dt.datetime.utcnow, onupdate=dt.datetime.utcnow)

    category = relationship("ProductCategory", back_populates="products")
    shelf = relationship("Shelf", back_populates="products")
    interactions = relationship("ProductInteraction", back_populates="product")
    attractiveness_scores = relationship("ProductAttractivenessScore", back_populates="product")

File: backend/models/product.js
const { DataTypes } = require('sequelize');
const { sequelize } = require('../config/db');

const Product = sequelize.define('Product', {
  id: {
    type: DataTypes.INTEGER,
    primaryKey: true,
    autoIncrement: true
  },
  product_id: {
    type: DataTypes.STRING(50),
    allowNull: false,
    unique: true
  },
  name: {
    type: DataTypes.STRING(200),
    allowNull: false
  },
  sku: {
    type: DataTypes.STRING(50),
    allowNull: false,
    unique: true
  },
  category: {
    type: DataTypes.STRING(100)
  },
  subcategory: {
    type: DataTypes.STRING(100)
  },
  brand: {
    type: DataTypes.STRING(100)
  },
  price: {
    type: DataTypes.FLOAT
  },
  size: {
    type: DataTypes.STRING(50)
  },
  color: {
    type: DataTypes.STRING(50)
  },
  shelf_id: {
    type: DataTypes.INTEGER,
    references: {
      model: 'shelves',
      key: 'id'
    }
  },
  shelf_level: {
    type: DataTypes.INTEGER
  },
  position_x: {
    type: DataTypes.FLOAT
  },
  position_y: {
    type: DataTypes.FLOAT
  },
  product_image_url: {
    type: DataTypes.STRING(500)
  },
  product_image_embedding: {
    type: DataTypes.JSONB,
    defaultValue: {}
  },
  status: {
    type: DataTypes.ENUM('active', 'inactive', 'discontinued'),
    defaultValue: 'active'
  },
  created_at: {
    type: DataTypes.DATE,
    defaultValue: DataTypes.NOW
  },
  updated_at: {
    type: DataTypes.DATE,
    defaultValue: DataTypes.NOW
  }
}, {
  tableName: 'products',
  timestamps: false
});

module.exports = Product;